src/systems/filecoin_vm/sysactors/init_actor.go

```diff
@@ -2,92 +2,134 @@ package sysactors
+	if a._isSingletonActor(codeCID) {
+		rt.Abort("cannot launch another actor of this type")
+	}
 
+	// Get the actor ID for this actor.
+	h, st := a.State(rt)
+	actorID := st._assignNextID()
+
+	// This generates a unique address for this actor that is stable across message
+	// reordering
+	// TODO: where do `creator` and `nonce` come from?
+	// TODO: CallSeqNum is not related to From -- it's related to Origin
+	// addr := rt.ComputeActorAddress(rt.Invocation().FromActor(), rt.Invocation().CallSeqNum())
+	addr := a._computeNewAddress(rt, actorID)
+
+	initBalance := rt.ValueSupplied()
+
+	// Set up the actor itself
+	actorState := &actor.ActorState_I{
+		CodeCID_: codeCID,
+		// State_:   nil, // TODO: do we need to init the state? probably not
+		Balance_:    initBalance,
+		CallSeqNum_: 0,
+	}
 
-		// Set up the actor itself
-		actorState := &actor.ActorState_I{
-			CodeCID_: codeCID,
-			// State_:   nil, // TODO: do we need to init the state? probably not
-			Balance_:    initBalance,
-			CallSeqNum_: 0,
-		}
+	stateCid := actor.StateCID(rt.IpldPut(actorState))
 
-		// The call to the actors constructor will set up the initial state
-		// from the given parameters, setting `actor.Head` to a new value when successfull.
-		// TODO: can constructors fail?
-		// TODO: this needs to be written such that the specific type Constructor is called.
-		//       right now actor.Constructor(p) calls the Actor type, not the concrete type.
-		// a.Constructor(params) // TODO: uncomment this.
+	// runtime.State().Storage().Set(actorID, actor)
 
-		// TODO: where is this VM.GlobalState?
-		// TODO: do we need this?
-		// runtime.State().Storage().Set(actorID, actor)
+	// Store the mappings of address to actor ID.
+	st.AddressMap()[addr] = actorID
+	st.IDMap()[actorID] = addr
 
-		// Store the mappings of address to actor ID.
-		state.AddressMap()[addr] = actorID
-		state.IDMap()[actorID] = addr
+	// TODO: adjust this to be proper state setting.
+	UpdateRelease(rt, h, st)
 
-		// TODO: adjust this to be proper state setting.
-		rt.State().StateTree().ActorStates()[addr] = actorState // atm it's nil
+	// TODO: can this fail?
+	rt.CreateActor(stateCid, addr, constructorParams)
 
-		return addr
-	*/
+	return rt.ValueReturn([]byte(addr.String()))
 }
 
-func (_ *InitActorCode_I) _assignNextID(state InitActorState) actor.ActorID {
-	stateI := state.Impl() // TODO: unwrapping like this is ugly.
-	actorID := stateI.NextID_
-	stateI.NextID_++
+func (s *InitActorState_I) _assignNextID() actor.ActorID {
+	actorID := s.NextID_
+	s.NextID_++
 	return actorID
 }
 
-func (_ *InitActorCode_I) GetActorIDForAddress(state InitActorState, address addr.Address) actor.ActorID {
-	return state.AddressMap()[address]
+func (_ *InitActorCode_I) _computeNewAddress(rt Runtime, id actor.ActorID) addr.Address {
+	// assign an address based on some randomness
+	// we use the current epoch, and the actor id. this should be a unique identifier,
```

Comment:
A message can easily end up in a different epoch during a re-org. I don't think this will produce a stable address.